src/cmd/services/m3coordinator/downsample/downsample_id.go

```diff
@@ -0,0 +1,330 @@
+	errNoMetricNameTag = errors.New("no metric name tag found")
+)
+
+// Ensure encodedTagsIterator implements id.SortedTagIterator
+var _ id.SortedTagIterator = &encodedTagsIterator{}
+
+// Ensure encodedTagsIterator implements id.ID
+var _ id.ID = &encodedTagsIterator{}
+
+type encodedTagsIterator struct {
+	tagDecoder serialize.TagDecoder
+	bytes      checked.Bytes
+	pool       *encodedTagsIteratorPool
+}
+
+func newEncodedTagsIterator(
+	tagDecoder serialize.TagDecoder,
+	pool *encodedTagsIteratorPool,
+) *encodedTagsIterator {
+	return &encodedTagsIterator{
+		tagDecoder: tagDecoder,
+		bytes:      checked.NewBytes(nil, nil),
+		pool:       pool,
+	}
+}
+
+// Reset resets the iterator.
+func (it *encodedTagsIterator) Reset(sortedTagPairs []byte) {
+	it.bytes.Reset(sortedTagPairs)
+	it.tagDecoder.Reset(it.bytes)
+}
+
+// Bytes returns the underlying bytes.
+func (it *encodedTagsIterator) Bytes() []byte {
+	return it.bytes.Bytes()
+}
+
+// TagValue returns the value for a tag value.
+func (it *encodedTagsIterator) TagValue(tagName []byte) ([]byte, bool) {
+	it.tagDecoder.Reset(it.bytes)
+	for it.Next() {
+		name, value := it.Current()
+		if bytes.Equal(tagName, name) {
+			return value, true
+		}
+	}
+	return nil, false
+}
+
+// Next returns true if there are more tag names and values.
+func (it *encodedTagsIterator) Next() bool {
+	return it.tagDecoder.Next()
+}
+
+// Current returns the current tag name and value.
+func (it *encodedTagsIterator) Current() ([]byte, []byte) {
+	tag := it.tagDecoder.Current()
+	return tag.Name.Bytes(), tag.Value.Bytes()
+}
+
+// Err returns any errors encountered.
+func (it *encodedTagsIterator) Err() error {
+	return it.tagDecoder.Err()
+}
+
+// Close closes the iterator.
+func (it *encodedTagsIterator) Close() {
+	it.bytes.Reset(nil)
+	it.tagDecoder.Reset(it.bytes)
+
+	if it.pool != nil {
+		it.pool.Put(it)
+	}
+}
+
+type encodedTagsIteratorPool struct {
+	tagDecoderPool serialize.TagDecoderPool
+	pool           pool.ObjectPool
+}
+
+func newEncodedTagsIteratorPool(
+	tagDecoderPool serialize.TagDecoderPool,
+	opts pool.ObjectPoolOptions,
+) *encodedTagsIteratorPool {
+	return &encodedTagsIteratorPool{
+		tagDecoderPool: tagDecoderPool,
+		pool:           pool.NewObjectPool(opts),
+	}
+}
+
+func (p *encodedTagsIteratorPool) Init() {
+	p.tagDecoderPool.Init()
+	p.pool.Init(func() interface{} {
+		return newEncodedTagsIterator(p.tagDecoderPool.Get(), p)
+	})
+}
+
+func (p *encodedTagsIteratorPool) Get() *encodedTagsIterator {
+	return p.pool.Get().(*encodedTagsIterator)
+}
+
+func (p *encodedTagsIteratorPool) Put(v *encodedTagsIterator) {
+	p.pool.Put(v)
+}
+
+func isRollupID(
+	tags []byte,
+	iteratorPool *encodedTagsIteratorPool,
+) bool {
+	iter := iteratorPool.Get()
+	iter.Reset(tags)
+
+	for iter.Next() {
+		name, value := iter.Current()
+		if bytes.Equal(name, rollupTagName) && bytes.Equal(value, rollupTagValue) {
+			return true
+		}
+	}
+	iter.Close()
+
+	return false
+}
+
+type rollupIDProvider struct {
```

Comment:
Can you add some comments explaining what this type is responsible for? It's hard to tell by just reading the code so far.

Reply:
Sure thing. (It just constructs a rollup ID when necessary, and can be pooled)